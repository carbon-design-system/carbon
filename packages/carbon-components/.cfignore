*
!Staticfile
!demo/
demo/*
!demo/index.html
!demo/demo.js
!demo/demo.min.js
!demo/demo.min.js.map
!demo/demo.css
!demo/code/
!demo/code/**
!demo/component/
!demo/component/**
