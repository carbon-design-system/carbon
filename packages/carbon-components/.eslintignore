demo/*.css
demo/*.map
demo/*.js
demo/js/prism.js
demo/hot
!demo/index.js
dist
examples/codesandbox
tests/a11y-results
tests/coverage
/es
/umd
/scripts
/css
/scss
/html
/docs/js
node_modules
