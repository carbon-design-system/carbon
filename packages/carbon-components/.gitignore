# Ignore built files and static folder
demo/*.css
demo/*.map
demo/*.js
demo/*.html
demo/code
demo/component
demo/hot
!demo/index.js
!demo/components.js
dist
docs/js

# Test reports
tests/a11y-results
tests/coverage
tests/axe/allHtml

# Built files
/scripts
/html
/scss

# a11y testing
.aat.yml

# Generated Sass files
src/globals/scss/vendor
