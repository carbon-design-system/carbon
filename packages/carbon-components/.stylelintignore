# packages/components
demo/**
tests/**
css/**
scss/**

