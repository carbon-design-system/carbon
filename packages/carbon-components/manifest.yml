---
applications:
- name: carbon-dev-environment
  memory: 64M
  buildpack: https://github.com/cloudfoundry/staticfile-buildpack.git
routes:
  - route: vanilla.carbondesignsystem.com
