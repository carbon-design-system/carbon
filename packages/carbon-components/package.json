{
  "name": "carbon-components",
  "description": "The Carbon Design System is IBMâ€™s open-source design system for products and experiences.",
  "version": "11.23.1",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/carbon-design-system/carbon.git",
    "directory": "packages/components"
  },
  "bugs": "https://github.com/carbon-design-system/carbon/issues",
  "homepage": "https://www.carbondesignsystem.com/",
  "files": [
    "scss/**/*",
    "index.scss"
  ],
  "keywords": [
    "carbon",
    "component",
    "components",
    "css",
    "ibm",
    "library",
    "pattern",
    "patterns",
    "sass",
    "scss",
    "style guide",
    "carbon-design-system",
    "react"
  ],
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "node tasks/build-styles.js",
    "postinstall": "carbon-telemetry collect --install"
  },
  "peerDependencies": {
    "sass": "^1.33.0"
  },
  "dependencies": {
    "@carbon/styles": "^1.23.1",
    "@carbon/telemetry": "0.1.0"
  },
  "devDependencies": {
    "@carbon/test-utils": "^10.26.0",
    "fs-extra": "^10.0.0",
    "rimraf": "^4.0.0",
    "sass": "^1.51.0"
  },
  "sideEffects": [
    "index.scss",
    "scss/**/*.scss",
    "scss/**/*.css"
  ]
}
