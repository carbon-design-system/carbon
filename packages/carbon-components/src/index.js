/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

// ====================//
// Imports and Exports //
// ====================//

// Base Elements & Components
// -------------
// - JavaScript classes for use with components and base-elements.
// - The following statements import classes from actual locations to
//   be consumed from this file instead of their actual locations.
export * from './globals/js/components';

export { default as settings } from './globals/js/settings';
