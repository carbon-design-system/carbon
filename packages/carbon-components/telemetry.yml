# yaml-language-server: $schema=https://unpkg.com/@ibm/telemetry-config-schema@v1/dist/config.schema.json

version: 1
projectId: fbc5603c-90ba-4d2f-b054-55108c9e06c0
endpoint: https://collector-prod.1am6wm210aow.us-south.codeengine.appdomain.cloud/v1/metrics
collect:
  npm:
    dependencies: null