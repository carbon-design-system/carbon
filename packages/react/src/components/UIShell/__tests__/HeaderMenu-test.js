/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import { HeaderMenu, HeaderMenuItem } from '../';
import { settings } from 'carbon-components';

const { prefix } = settings;

describe('HeaderMenu', () => {
  let mountNode;
  let mockProps;

  beforeEach(() => {
    mountNode = document.createElement('div');
    mockProps = {
      'aria-label': 'Accessibility label',
      className: 'custom-class',
      menuLinkName: 'test',
      // We use `ref` instead of `focusRef` becase `HeaderMenu` forwards the ref
      // to the underlying menu button
      ref: jest.fn(),
      tabIndex: -1,
    };

    document.body.appendChild(mountNode);
  });

  afterEach(() => {
    mountNode.parentNode.removeChild(mountNode);
  });

  it('should render', () => {
    const wrapper = mount(
      <HeaderMenu {...mockProps}>
        <HeaderMenuItem href="/a">A</HeaderMenuItem>
        <HeaderMenuItem href="/b">B</HeaderMenuItem>
        <HeaderMenuItem href="/c">C</HeaderMenuItem>
      </HeaderMenu>,
      {
        attachTo: mountNode,
      }
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render aria-label', () => {
    const wrapper = mount(
      <HeaderMenu {...mockProps}>
        <HeaderMenuItem href="/a">A</HeaderMenuItem>
        <HeaderMenuItem href="/b">B</HeaderMenuItem>
        <HeaderMenuItem href="/c">C</HeaderMenuItem>
      </HeaderMenu>,
      {
        attachTo: mountNode,
      }
    );
    const headerMenu = wrapper.childAt(0);
    const headerMenuText = headerMenu
      .find(`.${prefix}--header__menu-title`)
      .prop('aria-label');

    expect(headerMenuText).toMatch('Accessibility label');
  });

  it('should render content prop', () => {
    const menuContent = () => <p>Some other text</p>;
    const wrapper = mount(
      <HeaderMenu renderMenuContent={menuContent} {...mockProps}>
        <HeaderMenuItem href="/a">A</HeaderMenuItem>
        <HeaderMenuItem href="/b">B</HeaderMenuItem>
        <HeaderMenuItem href="/c">C</HeaderMenuItem>
      </HeaderMenu>,
      {
        attachTo: mountNode,
      }
    );

    const headerMenu = wrapper.childAt(0);
    const headerMenuAnchorChildText = headerMenu
      .find(`.${prefix}--header__menu-title`)
      .childAt(1)
      .text();
    const headerMenuText = headerMenu
      .find(`.${prefix}--header__menu-title`)
      .text();

    expect(headerMenuText).not.toMatch('Accessibility label');
    expect(headerMenuAnchorChildText).toMatch('Some other text');
  });

  describe('menu button interactions', () => {
    it('should should open and close', () => {
      const wrapper = mount(
        <HeaderMenu {...mockProps}>
          <HeaderMenuItem href="/a">A</HeaderMenuItem>
          <HeaderMenuItem href="/b">B</HeaderMenuItem>
          <HeaderMenuItem href="/c">C</HeaderMenuItem>
        </HeaderMenu>,
        {
          attachTo: mountNode,
        }
      );

      const headerMenu = wrapper.childAt(0);
      const headerInstance = headerMenu.instance();

      // Should start closed
      expect(headerInstance.state.expanded).toEqual(false);

      // Click should open
      headerMenu.simulate('click');
      expect(headerInstance.state.expanded).toEqual(true);

      // blur should close
      headerMenu.simulate('blur');
      expect(headerInstance.state.expanded).toEqual(false);

      // Get first link in the menu
      const menuLink = headerMenu.find('a').first();

      // After enter should open
      menuLink.simulate('keydown', { key: 'Enter', keyCode: 13, which: 13 });
      expect(headerInstance.state.expanded).toEqual(true);

      // After space should close
      menuLink.simulate('keydown', { key: 'Space', keyCode: 32, which: 32 });
      expect(headerInstance.state.expanded).toEqual(false);

      // After esc should close
      headerMenu.simulate('click');
      menuLink.simulate('keydown', { key: 'Escape', keyCode: 27, which: 27 });
      expect(headerInstance.state.expanded).toEqual(false);
    });
  });

  describe('menu button', () => {
    it.skip('should set the given ref on the menu button', () => {});

    it.skip('should open and close the menu when interacted with by a mouse', () => {});

    it.skip('should focus the first item in the menu if DOWN is pressed', () => {});

    it.skip('should focus the last item in the menu if UP is pressed', () => {});
  });

  describe('menu', () => {
    // https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-1/menubar-1.html#kbd2_label
    it.skip('should close the menu if RIGHT or LEFT are pressed', () => {});

    it.skip('should close the menu and set focus on the menu button if ESC is pressed', () => {});

    it.skip('should focus the first item if HOME is pressed', () => {});

    it.skip('should focus the last item if END is pressed', () => {});

    it.skip('should support movement with UP and DOWN keys', () => {});

    it.skip('should close the menu if blur occurs that is not in the menu or menubar', () => {});
  });
});
