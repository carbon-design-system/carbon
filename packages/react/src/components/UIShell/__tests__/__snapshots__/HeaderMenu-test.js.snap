// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderMenu should render 1`] = `
<HeaderMenu
  aria-label="Accessibility label"
  className="custom-class"
  menuLinkName="test"
  tabIndex={-1}
>
  <HeaderMenu
    aria-label="Accessibility label"
    className="custom-class"
    focusRef={
      [MockFunction] {
        "calls": Array [
          Array [
            <a
              aria-expanded="false"
              aria-haspopup="menu"
              aria-label="Accessibility label"
              class="bx--header__menu-item bx--header__menu-title"
              href="#"
              tabindex="0"
            >
              test
              <svg
                aria-hidden="true"
                class="bx--header__menu-arrow"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                />
              </svg>
            </a>,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    menuLinkName="test"
    renderMenuContent={[Function]}
    tabIndex={-1}
  >
    <li
      className="bx--header__submenu custom-class"
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
    >
      <a
        aria-expanded={false}
        aria-haspopup="menu"
        aria-label="Accessibility label"
        className="bx--header__menu-item bx--header__menu-title"
        href="#"
        onKeyDown={[Function]}
        tabIndex={0}
      >
        test
        <defaultRenderMenuContent>
          <ForwardRef(ChevronDown16)
            className="bx--header__menu-arrow"
          >
            <Icon
              className="bx--header__menu-arrow"
              fill="currentColor"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 16 16"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <svg
                aria-hidden={true}
                className="bx--header__menu-arrow"
                fill="currentColor"
                focusable="false"
                height={16}
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width={16}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                />
              </svg>
            </Icon>
          </ForwardRef(ChevronDown16)>
        </defaultRenderMenuContent>
      </a>
      <ul
        aria-label="Accessibility label"
        className="bx--header__menu"
      >
        <ForwardRef(HeaderMenuItem)
          href="/a"
          key=".0"
        >
          <li>
            <Link
              className="bx--header__menu-item"
              element="a"
              href="/a"
              tabIndex={0}
            >
              <a
                className="bx--header__menu-item"
                href="/a"
                tabIndex={0}
              >
                <span
                  className="bx--text-truncate--end"
                >
                  A
                </span>
              </a>
            </Link>
          </li>
        </ForwardRef(HeaderMenuItem)>
        <ForwardRef(HeaderMenuItem)
          href="/b"
          key=".1"
        >
          <li>
            <Link
              className="bx--header__menu-item"
              element="a"
              href="/b"
              tabIndex={0}
            >
              <a
                className="bx--header__menu-item"
                href="/b"
                tabIndex={0}
              >
                <span
                  className="bx--text-truncate--end"
                >
                  B
                </span>
              </a>
            </Link>
          </li>
        </ForwardRef(HeaderMenuItem)>
        <ForwardRef(HeaderMenuItem)
          href="/c"
          key=".2"
        >
          <li>
            <Link
              className="bx--header__menu-item"
              element="a"
              href="/c"
              tabIndex={0}
            >
              <a
                className="bx--header__menu-item"
                href="/c"
                tabIndex={0}
              >
                <span
                  className="bx--text-truncate--end"
                >
                  C
                </span>
              </a>
            </Link>
          </li>
        </ForwardRef(HeaderMenuItem)>
      </ul>
    </li>
  </HeaderMenu>
</HeaderMenu>
`;
